src/panoramic/cli/analytics.py

```diff
@@ -0,0 +1,199 @@
+
+logger = logging.getLogger(__name__)
+
+BLACKLISTED_ARGS = ['password']
+MINIMAL_FLUSH_DURATION = timedelta(minutes=10)
+MINIMAL_FLUSH_EVENTS = 10
+
+"""
+Usage: currently only way to track data is by using `write_command_event` after execution of each CLI command.
+If we decide to extend analytics with extra information, implement your own method like `write_request_event`
+that will track your data, but there is no need to call flush at the end.
+Data will be stored to events journal and flushed based on conditions set here.
+"""
+
+
+def write_command_event(name: str, group: str, params: Dict[str, Any], start_time: float, error: str = '') -> None:
+    """
+    Generate command execution context and store it in analytics events file as JSON line.
+    Truncate blacklisted parameters, like password.
+    Try to flush data to external source based on various conditions.
+    """
+    if not is_enabled():
+        return
+
+    for arg in BLACKLISTED_ARGS:
+        if arg in params:
+            del params[arg]
+
+    properties = {
+        'command': {
+            'name': ' '.join((group, name)),
+            'args': params,
+            'duration_seconds': time.time() - start_time,
+            'error': error,
+        },
+        'timestamp': time.time(),
+        'platform': platform.platform(),
+        'app_version': __version__,
+        'python_version': platform.python_version(),
+        'user_id': _get_user_id(),
+        'success': error == '',
+    }
+
+    # Store event into temporary JSON lines events file.
+    append_json_line(Paths.analytics_events_file(), properties)
+
+    # Do not crash the command because we are unable to connect with external source.
+    try:
+        _flush()
+    except Exception as e:
+        logger.debug("Analytics flush failed.", exc_info=e)
+
+
+def is_enabled() -> bool:
+    """Check if user opted out of usage metrics recording."""
+    if not config_is_enabled():
+        return False
+
+    config = read_config('analytics')
+    return config.get('enabled', False)
+
+
+def opt_in_command() -> None:
+    """CLI command to opt in to anonymous usage analytics."""
+    update_config('analytics', {'enabled': True})
+
+
+def opt_out_command() -> None:
+    """CLI command to opt out off anonymous usage analytics."""
+    update_config('analytics', {'enabled': False})
+
+    if Paths.analytics_dir().is_dir():
+        shutil.rmtree(Paths.analytics_dir())
+
+
+def show_tracking_id_command() -> None:
+    """CLI command to display users anonymous tracking id."""
+    echo_info(_get_user_id())
+
+
+def _get_user_id() -> str:
+    """Look for unique user ID that is used while storing events.
+    If no user ID was found, generate new id for user and store it in analytics metadata file.
+    """
+    metadata = _read_metadata()
+    if 'user_id' in metadata:
+        return metadata['user_id']
+
+    user_id = str(uuid.uuid4())
+    _update_metadata({'user_id': user_id})
+    return user_id
```

Comment:
Instead of this, can we use the client ID from the credentials? If the user hasn't put in credentials yet, let's just leave the user ID out of the report until the credentials are put in.

Reply:
Done